### PR TITLE
Actually migrate to tf 2.0

### PR DESCRIPTION
**Pull request by Tilps**
This is very much WIP, many many things unimplemented right now.

### CODE REVIEW

**Comment by Tilps** [comment]
This is almost done now.  Classical policy head network save/load still pending, but otherwise it should be a feature match for before tf2 conversion.

File: tf/chunkparser.py [hunk 1]
```diff
@@ -426,7 +426,7 @@ def test_tensorflow_parsing(self):
         best_q = best_q.reshape(batch_size, 3)
 
         # Pass it through tensorflow
-        with tf.Session() as sess:
+        with tf.compat.v1.Session() as sess:
```
Comment (Tilps) [inline tf/chunkparser.py]: I didn't convert this test.

File: tf/tfprocess.py [hunk 2]
```diff
@@ -128,191 +113,167 @@ def __init__(self, cfg):
...
+            self.swa_writer = tf.summary.create_file_writer(
                 os.path.join(os.getcwd(), "leelalogs/{}-swa-test".format(self.cfg['name'])))
-        self.histograms = [tf.summary.histogram(
-            weight.name, weight) for weight in self.weights]
-
-        self.init = tf.global_variables_initializer()
-        self.saver = tf.train.Saver()
-
-        self.session.run(self.init)
+        self.checkpoint = tf.train.Checkpoint(optimizer=self.orig_optimizer, model=self.model, global_step=self.global_step, swa_count=self.swa_count)
```
Comment (cn4750) [inline tf/tfprocess.py]: This line fails when `swa: false` is set since it is set to None and isn't a trackable object.

File: tf/train.py [hunk 3]
```diff
@@ -25,6 +25,7 @@
 import random
 import multiprocessing as mp
 import tensorflow as tf
+#tf.compat.v1.disable_v2_behavior()
```
Comment (Ttl) [inline tf/train.py]: This can be removed

File: tf/tfprocess.py [hunk 4]
```diff
@@ -523,134 +547,113 @@ def calculate_test_summaries(self, test_batches, steps):
...
+        permuted_tensors[-7] = all_tensors[-13]
+        permuted_tensors[-8] = all_tensors[-14]
+        permuted_tensors[-9] = all_tensors[-16]
+        permuted_tensors[-10] = all_tensors[-5]
+        permuted_tensors[-11] = all_tensors[-6]        
+        permuted_tensors[-12] = all_tensors[-7]        
+        permuted_tensors[-13] = all_tensors[-8]        
+        permuted_tensors[-14] = all_tensors[-9]        
+        permuted_tensors[-15] = all_tensors[-10]        
+        permuted_tensors[-16] = all_tensors[-15]
```
Comment (Ttl) [inline tf/tfprocess.py]: There is lot of trailing whitespace here.

File: tf/tfprocess.py [hunk 5]
```diff
@@ -523,134 +547,113 @@ def calculate_test_summaries(self, test_batches, steps):
...
+                # Biases, batchnorm etc
+                # pb expects every batch norm to have gammas, but not all of our
+                # batch norms have gammas, so manually add pretend gammas.
+                if 'beta:' in weights.name and not last_was_gamma:
+                    all_tensors.append(tf.ones_like(weights))
+                work_weights = weights.read_value()
+            all_tensors.append(work_weights)
+            last_was_gamma = 'gamma:' in weights.name
+
+        # HACK: model weights ordering is some kind of breadth first traversal,
```
Comment (Ttl) [inline tf/tfprocess.py]: This looks too fragile. Should probably rework the `net.py` to assign the weights in proto based on the tensor names instead. I'll try to look into if that looks doable.

File: tf/tfprocess.py (hunk 2 above)
Comment (Ttl) [reply tf/tfprocess.py]: Works when `self.swa_count` definition above is taken out of the if clause.

File: tf/tfprocess.py [hunk 6]
```diff
@@ -323,45 +284,90 @@ def replace_weights(self, new_weights):
...
+    @tf.function()
+    def read_weights(self):
+        return [w.read_value() for w in self.model.weights]
+
+    @tf.function()
+    def process_inner_loop(self):
+        print('tracing inner loop!')
+        x, y, z, q = next(self.train_iter)
+        with tf.GradientTape() as tape:
+            policy, value = self.model(x)
```
Comment (Ttl) [inline tf/tfprocess.py]: `self.model` call should have `training=True` to set batch norm to training mode.

File: tf/tfprocess.py [hunk 7]
```diff
@@ -439,76 +451,88 @@ def process(self, batch_size, test_batches, batch_splits=1):
...
         self.test_writer = true_test_writer
-        self.snap_restore()
+        for (old, w) in zip(backup, self.model.weights):
+            w.assign(old)
+
+    @tf.function()
+    def calculate_test_summaries_inner_loop(self):
+        print('tracing summaries inner loop!')
+        x, y, z, q = next(self.test_iter)
+        policy, value = self.model(x)
```
Comment (Ttl) [inline tf/tfprocess.py]: `training=False` here.

**Comment by Ttl** [comment]
With batch norm fixes I can restore old training using `net_to_model.py` with losses staying the same.

**Comment by Tilps** [comment]
Thanks for the catch on the batch norm not being right - all the examples of custom training loops in the tf2 documentation I found omit setting the training flag, but I've reviewed the tf code for keras model test_on_batch/train_on_batch in combination with graph model (like the kind we are using) and it seems they always explicitly pass the training flag as you suggest. So that is fixed.

File: tf/tfprocess.py (hunk 2 above)
Comment (Tilps) [reply tf/tfprocess.py]: Done.

File: tf/tfprocess.py [hunk 8]
```diff
@@ -323,45 +284,90 @@ def replace_weights(self, new_weights):
...
-            self.process(batch_size, test_batches, batch_splits=batch_splits)
+            self.process_v2(batch_size, test_batches, batch_splits=batch_splits)
+
+    @tf.function()
+    def read_weights(self):
+        return [w.read_value() for w in self.model.weights]
+
+    @tf.function()
+    def process_inner_loop(self):
+        print('tracing inner loop!')
```
[the remark this answers is not in the document]
Comment (Tilps) [reply tf/tfprocess.py]: Done.

File: tf/train.py (hunk 3 above)
Comment (Tilps) [reply tf/train.py]: Done.

File: tf/tfprocess.py [hunk 9]
```diff
@@ -439,76 +451,88 @@ def process(self, batch_size, test_batches, batch_splits=1):
...
-        self.calculate_test_summaries(test_batches, steps)
+        self.calculate_test_summaries_v2(test_batches, steps)
         self.test_writer = true_test_writer
-        self.snap_restore()
+        for (old, w) in zip(backup, self.model.weights):
+            w.assign(old)
+
+    @tf.function()
+    def calculate_test_summaries_inner_loop(self):
+        print('tracing summaries inner loop!')
```
[the remark this answers is not in the document]
Comment (Tilps) [reply tf/tfprocess.py]: Done.

File: tf/tfprocess.py (hunk 5 above)
Comment (Tilps) [reply tf/tfprocess.py]: I agree, but it was a lot easier to do this way than to do properly, so I'm going to stick with it for right now. If you want to give doing it right a shot, I'm happy :)

File: tf/tfprocess.py (hunk 4 above)
Comment (Tilps) [reply tf/tfprocess.py]: I think I fixed it...

File: tf/tfprocess.py (hunk 6 above)
Comment (Tilps) [reply tf/tfprocess.py]: Indeed - fixed.

File: tf/tfprocess.py (hunk 7 above)
Comment (Tilps) [reply tf/tfprocess.py]: Yeap.

**Comment by Tilps** [comment]
Going to submit with the experimental data loader unreviewed - its experimental.
Will follow up with more optimization and tweaks.
